File: app/src/androidTest/java/com/jbrunton/mymovies/fixtures/FragmentTestRule.java
package com.jbrunton.mymovies.fixtures;

import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.test.rule.ActivityTestRule;

public class FragmentTestRule<A extends AppCompatActivity, F extends Fragment> extends ActivityTestRule<A> {
    private static final String TAG = "FragmentTestRule";

    private final Class<F> fragmentClass;
    private final boolean launchFragment;
    private F fragment;

    public static <F extends Fragment> FragmentTestRule<?, F> create(Class<F> fragmentClass) {
        return new FragmentTestRule<>(AppCompatActivity.class, fragmentClass);
    }

    public static <F extends Fragment> FragmentTestRule<?, F> create(Class<F> fragmentClass, boolean initialTouchMode) {
        return new FragmentTestRule<>(AppCompatActivity.class, fragmentClass, initialTouchMode);
    }

    public static <F extends Fragment> FragmentTestRule<?, F> create(Class<F> fragmentClass, boolean initialTouchMode, boolean launchFragment) {
        return new FragmentTestRule<>(AppCompatActivity.class, fragmentClass, initialTouchMode, true, launchFragment);
    }

    public FragmentTestRule(Class<A> activityClass, Class<F> fragmentClass) {
        this(activityClass, fragmentClass, false);
    }

    public FragmentTestRule(Class<A> activityClass, Class<F> fragmentClass, boolean initialTouchMode) {
        this(activityClass, fragmentClass, initialTouchMode, true);
    }

    public FragmentTestRule(Class<A> activityClass, Class<F> fragmentClass, boolean initialTouchMode, boolean launchActivity) {
        this(activityClass, fragmentClass, initialTouchMode, launchActivity, true);
    }

    public FragmentTestRule(Class<A> activityClass, Class<F> fragmentClass, boolean initialTouchMode, boolean launchActivity, boolean launchFragment) {
        super(activityClass, initialTouchMode, launchActivity);
        this.fragmentClass = fragmentClass;
        this.launchFragment = launchFragment;
    }

    @Override
    protected void afterActivityLaunched() {
        if (launchFragment) {
            launchFragment(createFragment());
        }
    }

    public void launchFragment(final F fragment) {
        try {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    final F fragment2 = fragment == null ? createFragment() : fragment;
                    FragmentTestRule.this.fragment = fragment2;
                    getActivity().getSupportFragmentManager()
                            .beginTransaction()
                            .replace(android.R.id.content, fragment2)
                            .commitNow();
                }
            });
        } catch (Throwable throwable) {
            throw new RuntimeException(throwable);
        }
    }

    protected F createFragment() {
        try {
            return fragmentClass.newInstance();
        } catch (InstantiationException e) {
            throw new AssertionError(String.format("%s: Could not insert %s into %s: %s",
                    getClass().getSimpleName(),
                    fragmentClass.getSimpleName(),
                    getActivity().getClass().getSimpleName(),
                    e.getMessage()));
        } catch (IllegalAccessException e) {
            throw new AssertionError(String.format("%s: Could not insert %s into %s: %s",
                    getClass().getSimpleName(),
                    fragmentClass.getSimpleName(),
                    getActivity().getClass().getSimpleName(),
                    e.getMessage()));
        }
    }

    public F getFragment() {
        if (fragment == null) {
            Log.w(TAG, "Fragment wasn't created yet");
        }
        return fragment;
    }
}

File: app/src/androidTest/java/com/jbrunton/mymovies/fixtures/FragmentTestRule.kt
package com.jbrunton.mymovies.fixtures

import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.test.rule.ActivityTestRule

open class FragmentTestRule<A : AppCompatActivity, F : Fragment>(
        activityClass: Class<A>,
        private val fragmentClass: Class<F>,
        initialTouchMode: Boolean,
        launchActivity: Boolean,
        private val launchFragment: Boolean
) : ActivityTestRule<A>(activityClass, initialTouchMode, launchActivity) {
    lateinit var fragment: F

    override fun afterActivityLaunched() {
        if (launchFragment) {
            launchFragment(createFragment())
        }
    }

    fun launchFragment(fragment: F?) {
        try {
            runOnUiThread {
                val instance = fragment ?: createFragment()
                this@FragmentTestRule.fragment = instance
                activity.supportFragmentManager
                        .beginTransaction()
                        .replace(android.R.id.content, instance)
                        .commitNow()
            }
        } catch (throwable: Throwable) {
            throw RuntimeException(throwable)
        }

    }

    protected open fun createFragment(): F {
        try {
            return fragmentClass.newInstance()
        } catch (e: InstantiationException) {
            throw AssertionError(String.format("%s: Could not insert %s into %s: %s",
                    javaClass.simpleName,
                    fragmentClass.simpleName,
                    activity.javaClass.simpleName,
                    e.message))
        } catch (e: IllegalAccessException) {
            throw AssertionError(String.format("%s: Could not insert %s into %s: %s",
                    javaClass.simpleName,
                    fragmentClass.simpleName,
                    activity.javaClass.simpleName,
                    e.message))
        }

    }

    companion object {
        fun <F : Fragment> create(
                fragmentClass: Class<F>,
                initialTouchMode: Boolean = false,
                launchActivity: Boolean = true,
                launchFragment: Boolean = true): FragmentTestRule<*, F> {
            return FragmentTestRule(AppCompatActivity::class.java, fragmentClass, initialTouchMode, launchActivity, launchFragment)
        }
    }
}